src/main/ApplicationImpl.cpp

```diff
@@ -549,27 +550,224 @@ ApplicationImpl::joinAllThreads()
+
+    // The "scphistory" stores ledger sequence numbers as INTs.
+    if (startLedgerSeq >= std::numeric_limits<int32_t>::max())
+    {
+        throw std::invalid_argument(fmt::format(
+            FMT_STRING("Manually closed ledger sequence number ({}) can not be "
+                       "increased without overflowing database entry"),
+            startLedgerSeq));
+    }
+
+    auto const nextLedgerSeq = startLedgerSeq + 1;
+
+    if (userSpecifiedSeqNum)
+    {
+        if (*userSpecifiedSeqNum > std::numeric_limits<int32_t>::max())
+        {
+            throw std::invalid_argument(fmt::format(
+                FMT_STRING("Manual close ledger sequence number {} too large "
+                           "for database (maximum {})"),
+                *userSpecifiedSeqNum, std::numeric_limits<int32_t>::max()));
+        }
+
+        if (*userSpecifiedSeqNum <= startLedgerSeq)
+        {
+            throw std::invalid_argument(fmt::format(
+                FMT_STRING("Invalid manual close ledger sequence number {} "
+                           "(must exceed current sequence number {})"),
+                *userSpecifiedSeqNum, startLedgerSeq));
+        }
+    }
+
+    return userSpecifiedSeqNum ? *userSpecifiedSeqNum : nextLedgerSeq;
+}
+
+void
+ApplicationImpl::setManualCloseVirtualTime(
+    optional<TimePoint> const& userSpecifiedCloseTime)
+{
+    TimePoint const maxSystemTimePointSeconds =
+        std::chrono::duration_cast<std::chrono::seconds>(
+            VirtualClock::system_time_point::duration::max())
+            .count();
+    TimePoint const maxTimeT = std::numeric_limits<std::time_t>::max();
+    TimePoint const maxCloseTime =
+        std::min(maxSystemTimePointSeconds, maxTimeT) -
+        Herder::MAX_TIME_SLIP_SECONDS.count();
+
+    TimePoint const lastCloseTime = getLedgerManager()
+                                        .getLastClosedLedgerHeader()
+                                        .header.scpValue.closeTime;
+    if (lastCloseTime >= maxCloseTime)
+    {
+        throw std::invalid_argument(
+            fmt::format(FMT_STRING("Close time already at max ({}); can not "
+                                   "close further ledgers"),
+                        maxCloseTime));
+    }
+
+    auto systemNow = getClock().system_now();
+    TimePoint secondsSinceEpoch =
+        std::chrono::duration_cast<std::chrono::seconds>(
+            systemNow.time_since_epoch())
+            .count();
+    if (secondsSinceEpoch > maxCloseTime)
+    {
+        throw std::invalid_argument(
+            fmt::format(FMT_STRING("System time already past max ({}); can not "
+                                   "close further ledgers"),
+                        maxCloseTime));
+    }
+
+    uint64_t const now = VirtualClock::to_time_t(systemNow);
+    uint64_t nextCloseTime = std::max(now, lastCloseTime + 1);
+
+    if (userSpecifiedCloseTime)
+    {
+        if (*userSpecifiedCloseTime > maxCloseTime)
```

Comment:
I think we should just move those checks before calling `setCurrentVirtualTime` (and tweak the wording a bit as it's not necessarily provided by the user) - that way we perform the checks on the validity of `nextCloseTime` regardless of default behavior or not.

Reply:
I think that has also happened as part of 4932dab679e9f802c211ee35dd8d1c83655bb6e5.